contrib/educational_decoder/zstd_decompress.c

```diff
@@ -0,0 +1,2168 @@
+        }
+        size = read_bits_LE(streams->src, 20, 4);
+        streams->src += 3;
+        streams->src_len -= 3;
+        break;
+    default:
+        // Impossible
+        size = -1;
+    }
+
+    if (size > MAX_LITERALS_SIZE) {
+        CORRUPTION();
+    }
+
+    *literals = malloc(size);
+    if (!*literals) {
+        BAD_ALLOC();
+    }
+
+    switch (block_type) {
+    case 0:
+        // Raw data
+        if (size > streams->src_len) {
+            INP_SIZE();
+        }
+        memcpy(*literals, streams->src, size);
+        streams->src += size;
+        streams->src_len -= size;
+        break;
+    case 1:
+        // Single repeated byte
+        if (1 > streams->src_len) {
+            INP_SIZE();
+        }
+        memset(*literals, streams->src[0], size);
+        streams->src += 1;
+        streams->src_len -= 1;
+        break;
+    }
+
+    return size;
+}
+
+/// Decodes Huffman compressed literals
+static size_t decode_literals_compressed(io_streams_t *const streams,
+                                         frame_context_t *const ctx,
+                                         u8 **const literals,
+                                         const int block_type,
+                                         const int size_format) {
+    size_t regenerated_size, compressed_size;
+    // Only size_format=0 has 1 stream, so default to 4
+    int num_streams = 4;
+    switch (size_format) {
+    case 0:
+        num_streams = 1;
+    // Fall through as it has the same size format
+    case 1:
+        if (streams->src_len < 3) {
+            INP_SIZE();
+        }
+        regenerated_size = read_bits_LE(streams->src, 10, 4);
+        compressed_size = read_bits_LE(streams->src, 10, 14);
+        streams->src += 3;
+        streams->src_len -= 3;
+        break;
+    case 2:
+        if (streams->src_len < 4) {
+            INP_SIZE();
+        }
+        regenerated_size = read_bits_LE(streams->src, 14, 4);
+        compressed_size = read_bits_LE(streams->src, 14, 18);
+        streams->src += 4;
+        streams->src_len -= 4;
+        break;
+    case 3:
+        if (streams->src_len < 5) {
+            INP_SIZE();
+        }
+        regenerated_size = read_bits_LE(streams->src, 18, 4);
+        compressed_size = read_bits_LE(streams->src, 18, 22);
+        streams->src += 5;
+        streams->src_len -= 5;
+        break;
+    default:
+        // Impossible
+        compressed_size = regenerated_size = -1;
+    }
+    if (regenerated_size > MAX_LITERALS_SIZE ||
+        compressed_size > regenerated_size) {
+        CORRUPTION();
+    }
+
+    if (compressed_size > streams->src_len) {
+        INP_SIZE();
+    }
+
+    *literals = malloc(regenerated_size);
+    if (!*literals) {
+        BAD_ALLOC();
+    }
+
+    if (block_type == 2) {
+        // Decode provided Huffman table
+
+        HUF_free_dtable(&ctx->literals_dtable);
+        const size_t size = decode_huf_table(streams->src, compressed_size,
```

Comment:
I'm not sure if it makes sense to pass `streams` to `HUF` and `FSE` or not. Thats up to you, but if you pass a raw pointer, you should do something like this (modulo function names).

```c
const u8 *const src = streams_get_src_bytes(streams, compressed_size); // does not update state but does bounds checks
size_t const size = decode_huf_tables();
streams_advance_src(streams, size); // updates state
```